### PR TITLE
Refactor Maximum Subarray Implementation

### PR DESCRIPTION
**Pull request by TruongNhanNguyen**
## Checklist:

- [x] I ran bellow commands using the latest version of **rust nightly**.
- [x] I ran `cargo clippy --all -- -D warnings` just before my last commit and fixed any issue that was found.
- [x] I ran `cargo fmt` just before my last commit.
- [x] I ran `cargo test` just before my last commit and all tests passed.
- [x] I added my algorithm to the corresponding `mod.rs` file within its own folder, and in any parent folder(s).
- [x] I added my algorithm to `DIRECTORY.md` with the correct link.
- [x] I checked `COUNTRIBUTING.md` and my code follows its guidelines.

### CODE REVIEW

**Comment by TruongNhanNguyen** [comment]
Can you have a look and review this, @vil02?

File: src/dynamic_programming/maximum_subarray.rs [hunk 1]
```diff
@@ -1,62 +1,73 @@
-/// ## maximum subarray via Dynamic Programming
+//! This module provides a function to find the subarray with the largest sum
```
Comment (vil02) [inline src/dynamic_programming/maximum_subarray.rs:1]: The current implementation returns the sum of such subarray.

File: src/dynamic_programming/maximum_subarray.rs [hunk 2]
```diff
@@ -1,62 +1,73 @@
...
-    result
+    let mut max_sum = vec![0; array.len()];
+    max_sum[0] = array[0];
+
+    array.iter().enumerate().skip(1).for_each(|(i, &x)| {
+        max_sum[i] = (max_sum[i - 1] + x).max(x);
+    });
+
+    *max_sum.iter().max().unwrap()
 }
```
Comment (vil02) [inline src/dynamic_programming/maximum_subarray.rs:35]: I think you can optimize it to 
```rust
pub fn maximum_subarray(array: &[isize]) -> isize {
    if array.is_empty() {
        return 0;
    }

    let mut cur_sum = array[0];
    let mut max_sum = cur_sum;

    for &x in &array[1..] {
        cur_sum = (cur_sum + x).max(x);
        max_sum = max_sum.max(cur_sum);
    }

    max_sum
}
```
Above has $O(1)$ space complexity.

File: src/dynamic_programming/maximum_subarray.rs [hunk 3]
```diff
@@ -1,62 +1,73 @@
...
-        let array = vec![-6];
-        assert_eq!(maximum_subarray(&array), -6);
+    maximum_subarray_tests! {
+        test_all_non_negative: (vec![1, 0, 5, 8], 14),
+        test_all_negative: (vec![-3, -1, -8, -2], -1),
+        test_mixed_negative_and_positive: (vec![-4, 3, -2, 5, -8], 6),
+        test_single_element_positive: (vec![6], 6),
+        test_single_element_negative: (vec![-6], -6),
+        test_mixed_elements: (vec![-2, 1, -3, 4, -1, 2, 1, -5, 4], 6),
+        test_empty_array: (vec![], 0),
```
Comment (vil02) [inline src/dynamic_programming/maximum_subarray.rs:60]: According to [line 5](https://github.com/TheAlgorithms/Rust/blob/18fe64aeb95674f96aef5de30f46771d4a5188ee/src/dynamic_programming/maximum_subarray.rs#L5), this should return some error.

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/dynamic_programming/maximum_subarray.rs (hunk 2 above)
Comment (TruongNhanNguyen) [reply src/dynamic_programming/maximum_subarray.rs:35]: Good point here

File: src/dynamic_programming/maximum_subarray.rs (hunk 3 above)
Comment (TruongNhanNguyen) [reply src/dynamic_programming/maximum_subarray.rs]: Added custom error type

**Review by vil02 — APPROVED** [review]

**Pull request merged**